### ISSUE TITLE
Choose a serialized form for graphs

### ISSUE THREAD

**Issue by wchargin** [sourcecred/sourcecred#65]
Graphs should have a serialized form so that we can use Jest’s snapshot testing. Two options seem reasonable to me, so I implemented them.

On branch `wchargin-graph-json-arrays`, we have
```js
export type GraphJSON = {|
  +nodes: Node<mixed>[],
  +edges: Edge<mixed>[],
|};
```
which is a completely non-redundant format, but for which it is possible to have multiple JSON representations of logically-equal graphs, because the arrays are really sets and so are invariant under permutation. If we choose this method, we’ll provide an additional `toCanonicalJSON` method that sorts the `nodes` and `edges` by their addresses.

On branch `wchargin-graph-json-objects`, we have
```js
export type GraphJSON = {|
  +nodes: {[nodeAddress: string]: Node<mixed>},
  +edges: {[edgeAddress: string]: Edge<mixed>},
|};
```
which satisfies that if `a.equals(b)` then `deepEqual(a.toJSON(), b.toJSON())` at the cost of a bit of redundancy and potential for mismatch: i.e., we should require that `addressToString(json.nodes[key].address) === key` and likewise for edges, and need to guard against bad inputs. On the other hand, this form is somewhat more useful to external clients that want to directly consume the JSON.

On our `advancedMealGraph`, the latter is about 15% bigger in terms of file size:
```shell
$ filename=src/backend/__snapshots__/graph.test.js.snap
$ git show origin/wchargin-graph-json-arrays:"${filename}" | wc -c
5602
$ git show origin/wchargin-graph-json-objects:"${filename}" | wc -c
6476
```
This factor will increase as addresses get longer in proportion to the payloads, and indeed for our `advancedMealGraph` the addresses are reasonably short.

It’s worth reading the diffs at 574f916d0ecb1f167a6bd013de6f58d22625e2da (arrays) and 60771c56e257a398d23632ed64391f3df05439e0 (objects). In particular, note that using the objects form gets us free interop with Jest and `JSON.stringify`.

I’m inclined to go with the objects solution due to its simplicity in not requiring an additional canonical form. This has already paid off in the Jest and stringify interops, and it seems totally plausible that similar dividends will continue to arise down the road.

Once we come to a consensus, I can PR one of the two branches.

cc @dandelionmane

**Comment by wchargin** [comment]
> This factor will increase as addresses get longer in proportion to the payloads, and indeed for our `advancedMealGraph` the addresses are reasonably short.

A thought: plugins, or the Graph class itself, could choose to SHA-3 (or similar) the string-form keys to save space in the address table, if desired. This could have significant space savings for keys that are `JSON.stringify(someComplexStructure)`, at the cost of some time.

**Comment by dandelionmane** [comment]
Given these two options, I prefer the object approach. It has better semantics: the way we think of the graph is that it actually *is* a map in which nodes have unique keys, and so storing it this way matches what the data structure actually means.

However, it is redundant, because we essentially have a `Map k (k, v)`.

An aside on types. We had an offline discussion about how we want to set up our `Node` type: do we want to express subtyping by saying `Node`s have a payload with subtype, or `Node`s are extended to the subtype? I.e.:

Payload Approach:
```ts
type Node<T> {
  address: Address,
  payload: T,
}

type IssuePayload {
  title: string,
}

type IssueNode = Node<IssuePayload>
```

Extension Approach:

```ts
type Node {
  address: Address,
}

type IssueNode extends Node {
  title: string,
}
```

We decided to go with the payload approach; both because it's considered best practice to prefer composition over inheritance, and because it creates a data structure that explicitly reflects how our data is composed (ie: everything from plugin is under payload/, rather than plugin data is mixed in at the same level).

However, when you consider the Edge type:

```ts
type Edge<T> {
  address: Address,
  src: Address,
  dst: Address,
  payload: T,
}
```

It starts to look like both Edge and Node are implicitly inheriting from an undeclared base type Addressable:

```ts
type Addressable {
  address: Address,
}
```

While writing test code, we've actually reified the Addressable type a few times, which suggests that it's a useful concept. 

Would it be useful to factor out `Addressable`s as a payload-bearing-type like we did with `Node` and `Edge`? E.g. 

```ts
type Addressable<T> {
  address: Address,
  payload: T,
}
```

Except really I think we can more clearly express ourselves by understanding that the semantics we want from addresses are really just key-value semantics, where the keys are arbitrary JSON-serializable objects, and as an implementation detail, we can easily generate a map over Addressables by JSON-stringifying the key.

```ts
type Addressable<T> {
  key: JSONSerialiazable,
  value: T,
}
```
If we factor things like this, I think there's a very natural answer to "how should AddressableMaps be serialized": when JSON serializing, we stringify the keys, and elide them from the objects to avoid redundant encoding. When deserializing, we parse the keys into Address objects so that consumers can read the addresses as expected. 

Basically, this whole structure is an abstraction over a bug, where the bug is that you can't use JSON objects as keys in js maps.

**Comment by wchargin** [comment]
> It starts to look like both Edge and Node are implicitly inheriting from an undeclared base type Addressable: […] While writing test code, we've actually reified the Addressable type a few times, which suggests that it's a useful concept.

Agreed. It could also make sense to make it an interface:
```js
interface Addressable {
  +address: Address;
}
```
(This appeals to me in part because we’ve had lots of trouble getting Flow to work as expected with intersection types and similar. Also, semantically it _is_ an interface.)
   
> Would it be useful to factor out `Addressable`s as a payload-bearing-type like we did with Node and Edge? E.g. 
> ```js
> type Addressable<T> {
>   address: Address,
>   payload: T,
> }
> ```

I don’t think so, because this doesn’t work nicely for edges. What happens to `src` and `dst`? You could put them into the payload, so that
```js
type Edge<T> = Addressable<{| +src: Address, +dst: Address, +realPayload: T |}>
```
but that’s quite ugly, IMO: `edge.payload.realPayload` is ungainly, and `edge.payload.src` doesn’t make sense. Semantically, `src` and `dst` are _not_ part of the payload of the edge. They are metadata, like the address. But if we put them at top level, then `Edge<T>` adds additional properties onto `Addressable<T>`. This is weird, because the structure of this version of `Addressable<T>` strongly suggests that an object `x: Addressable<T>` is determined only by its `address` and its `payload`. For instance, given this interface, I would feel comfortable writing an equality function:
```js
function equals<T>(x: Addressable<T>, y: Addressable<T>) {
  return deepEqual(x.address, y.address) && deepEqual(x.payload, y.payload);
}
```
which would fail on edges.

I also don’t really see any advantage to including the payload in the `Addressable` interface, so I think that the simple `interface Addressable { +address: Address; }` is the way to go.

You also propose an alternate interface where the `Addressable`’s key is an arbitrary `JSONSerializable`. This idea sounds reasonable to me, but I would also consider putting the `JSONSerializable` on the `id` field of the `Address` type, so that all addresses are still required to have a repository and plugin name:
```js
type Address = {|
  +repositoryName: string,
  +pluginName: string,
  +id: JSONSerializable,
|};
```
This is structure that we do want in the graph, so we have to encode it somewhere. And putting it in the address is the only reasonable way that I can see to do it.

What do you think?

> Basically, this whole structure is an abstraction over a bug, where the bug is that you can't use JSON objects as keys in js maps.

…yes, and that ES6 maps are keyed against object identity. :-)

**Comment by dandelionmane** [comment]
> I also don’t really see any advantage to including the payload in the Addressable interface, so I think that the simple interface Addressable { +address: Address; } is the way to go.

Agreed.

> You also propose an alternate interface where the Addressable’s key is an arbitrary JSONSerializable. This idea sounds reasonable to me, but I would also consider putting the JSONSerializable on the id field of the Address type, so that all addresses are still required to have a repository and plugin name:
>
>```
>type Address = {|
>  +repositoryName: string,
>  +pluginName: string,
>  +id: JSONSerializable,
>|};
>```
>
> This is structure that we do want in the graph, so we have to encode it somewhere. And putting it in the address is the only reasonable way that I can see to do it.

Agreed.

I also suggested that in the Object serialization approach, we can use the JSON serialization of the Address as the object keys, and elide them from the Addressables. (And on deserialization we put them back.) What do you think of that?

**Comment by dandelionmane** [comment]
Another point in favor of the explicit Address struct: Per [my perf testing](https://github.com/sourcecred/sourcecred/issues/66#issuecomment-370321398), that approach is a lot faster - assuming that the JSON stringification is handled by the plugins, so that generic graph code just operates on string IDs.

**Comment by wchargin** [comment]
> I also suggested that in the Object serialization approach, we can use the JSON serialization of the Address as the object keys, and elide them from the Addressables. (And on deserialization we put them back.) What do you think of that?

It seems reasonable in practice, and I'm open to it. In practice: this is a realm of Flow that _should_ work just fine, though I would not be surprised if we quickly run into corner cases.

For instance, we’ll need a type like
```js
type SansAddress<T: Addressable> = $Exact<$Diff<T, {+address: Address}>>;
```
(without the `$Exact`, we have `T <: SansAddress<T>`, which is not correct) so that we can write
```js
export type AddressMapJSON<T> = {[serializedAddress: string]: SansAddress<T>};
```
This is mostly well and good\*. The type error messages, however, are not great. For instance, with the type
```js
type Node<T> = {
  +address: Address,
  +payload: T,
};
```
…I can produce [two statements][bad-errors] that have real type errors, which Flow detects but surfaces with the following inordinately unhelpful messages:
  - “Cannot get `bareNode.address` because property `address` is missing in `Node` [1].”
  - “Cannot assign `node` to `bareNode` because property `address` is missing in `Node` [1] but exists in `Node` [2].”
  - “Cannot assign `node` to `bareNode` because property `payload` is read-only in `Node` [1] but writable in `Node` [2].”

The first two are *extremely* understandabad. The last one…I don’t even know.

[bad-errors]: https://flow.org/try/#0PTAEAEDMBsHsHcBQiCmAPADrATgF1LgJ4YqgBSAygPIByFK2AlgIbSMBezARtKQLyJQoAD6gA3gG0AzriYA7AOYBdAFzlqdBizaceKAL6CR62vSasO3XhKVHRM+Qrug5AVwC2XBs66xYvZjlnN2hoAG5kdCw8AmJSAEEAE0TsFCkpUD5xYSMAalSsKUZcHEIaZncUNQdGRQAaPIxoVwVa8srq2VqFBqFcxkS1SlMtC11eBuF9CMRa3AZIZgBjBOTU9KtSMTzmNbSpNSSU-YjDVEwcfCISUApAqSP1qQAeABVDvY29AD5M0AASACiaGWuGe-wAIoxIJA3nVxLldsd0h9kVJ9N9vhEopdYjdHvsALLMDDDGhvVFPTa-LKSKSjHQoRIElGgGqKVS3e4sl6vb7TZDXUg0WCJFBvGnZHafA6gHm9UC5DDMQhwXZqV6TAWISCuORLXCMWByNldDAUuUy6kAClwGoAlGo7nIHjKJeIjEtjTJQOtXNB8LSAHQh3ACoRi3jzX1pf24INIp4RISpXCubAm61+gNqQKEe2nZBel34OSiqqgEVi55uTwMSXbISJ-ZqRtCGOFYqldoVgBEUlg6ZWS1SiWAA6HKBHTN7CqETRabQqfcY6VcKAAtAA2ACss6MQgGal7XAqvlg7F4QYAVhgFPuhPoFcrVbB1aAAIwAZga2qjoFPVIqwrZ1XTRZ5gJrDwvGwTE-hqDBrTLMUC0QQCUGAhMZTCIQQBjVhoEIUAlkCAByfBlhWdI0OYIDyz+ZCUBw0A8KkAALQdoESADSAAAwQpDy3tXjQFqGQUF2ZBgAAKkQa17UyX4JWtNA1EgvlHS5F0eQg8sJRpX40AiYsfSkRinW5N1INrGC4NpIxm1ZNsUxQTsSmwMpl2PCdsGHUdx0HXyp1HB92wXVo5B7Y9Vykddtz3OdRMGUATzPPxLxQG871Cp8jBfNVku-X8Imk4AgA

Valid code appears to type-check correctly, though. [These tests][tests] pass and yield the following snapshot:
```json
{
    "{\"repositoryName\":\"sourcecred/sourcecred\",\"pluginName\":\"issue65\",\"id\":{\"en\":\"one\"}}": {
        "payload": "uno"
    },
    "{\"repositoryName\":\"sourcecred/sourcecred\",\"pluginName\":\"issue65\",\"id\":{\"en\":\"two\"}}": {
        "payload": "dos"
    }
}
```

[tests]: https://gist.github.com/wchargin/745ef7c3b8991443cf1ae5bbcf60b02b

So I support it in principle and also happen to think that it’s tractable—good enough for me.

One caveat to my approval: this is strictly an optimization, and its implementation trades off smaller space for more complicated code, potentially more complicated interop (external consumers will be confused to see missing addresses), and potentially longer time (though potentially faster here, too, due to serializing less). Before making any final decisions, we should look at the actual space and time differences on some real data!

\* It may behave weirdly with exact types. I’m not sure what exactly is going on.

**Comment by wchargin** [comment]
> where the keys are arbitrary JSON-serializable objects

After a bit more thinking and prototyping, I’m less sold on this. I think that it forces a lot more serialization and deserialization than is really necessary. Any time that we need to refer to an address opaquely, we’ll have to serialize and deserialize the plugin-specific `id`—which may be arbitrarily large. For instance, a breadth-first search requires at least one roundtrip serialization for each node in the connected component of the source (and with our APIs we’re looking at two or three). If `Address` has `id: string`, then no such computation is performed: a plugin only parses IDs when their contents are relevant.

Using `id: string` of course _allows_ plugins to serialize the data from JSON themselves, but does not require them to. Using `id: JSONSerializable` would be convenient for the standard use case, and promote good defaults, but it _may_ be the case that it precludes efficient implementations.

Note that this is compatible with your other suggestion that “we can use the JSON serialization of the Address as the object keys, and elide them from the Addressables. (And on deserialization we put them back.)”

We’ve spent a bunch of time talking about this, so here’s what I suggest: I’ll implement an address map against this spec, just so that we have something to work from, and then we can remain open to changing it later—either in response to perf requirements or because we decide that another approach is semantically better. Sound good?

**Comment by dandelionmane** [comment]
Sounds good. As you say, using `id: string` allows plugins to layer in additional information, while restricting the performance cost of maintaining that additional information to the scope of the plugin that actually knows how to use the info. Generic algorithms operating at the level of a ContributionGraph shouldn't be expected to have insight into what the plugin specific ID means, so therefore they shouldn't pay the associated serialization cost.

**Comment by wchargin** [comment]
> Generic algorithms operating at the level of a ContributionGraph shouldn't be expected to have insight into what the plugin specific ID means, so therefore they shouldn't pay the associated serialization cost.

Yep, exactly.

**Comment by wchargin** [comment]
Implemented in #67, #68, #69.
